### ISSUE TITLE
"slice" expressions with number input within "==" expression gives unexpected result

### ISSUE THREAD

**Issue by pathmapper** [mapbox/mapbox-gl-js#9818]
**mapbox-gl-js version**: v1.11.0

**browser**: Firefox, Chrome

### Steps to Trigger Behavior

Use a [slice expression](https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#slice) with a number input like this:
`["case", ["==", ["slice", {number}, 0, 1], "1"], "True", "Fallback"]`

### Link to Demonstration

https://jsbin.com/calonadica/1/edit?html,output

### Expected Behavior

There is an error logged in the console and the evaluation fails or the fallback value is selected.

### Actual Behavior

The fallback value is not selected and there is only the following warning in the console:

`Expected first argument to be of type array or string, but found number instead.`

**Comment by karimnaaji** [comment]
Thanks for the example, I just tested that and it does look like a bug.

**Comment by ryanhamley** [comment]
I don't think this is a bug. The `slice` expression uses the `slice` method internally which only applies to strings and arrays. We could detect numbers, convert them to a string, slice them and convert them back to a number but that's more of a feature request to expand the functionality of the expression.

**Comment by pathmapper** [comment]
Thanks for your feedback.

> The slice expression uses the slice method internally which only applies to strings and arrays.

Sure, I opened the issue because of the current evaluation of the `==` expression.

What's currently the output of `["slice", {number}, 0, 1]` ?

The possible _results_ of a `==` expression according to the [docs](https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#==) are:
- `true`
- `false`
- parse error 

For `["==", ["slice", {number}, 0, 1], "1"]` it's definately not `true`. So I would expect either `false` or a parse error, but it's none of them (supposing that a parse error would be logged in the console).

**Comment by ryanhamley** [comment]
The `slice` expression is throwing a parsing error in this case. It sounds like your expectation is that `==` should handle an error in a sub-expression and either return `false` or it's own parsing error. I see the argument there but I wonder if it would do more harm than good to catch sub-expression errors. Alternatively, the `==` expression could just pass on the sub-expression error. I definitely don't think returning `false` would be a helpful response as it would be very hard to debug.

**Comment by pathmapper** [comment]
> It sounds like your expectation is that == should handle an error in a sub-expression and either return false or it's own parsing error.

Agree that `false` wouldn't be helpful. I would prefer some kind of error, to make debugging easier. Currently a nested `slice` expression could fail and you won't see an error. 

> The slice expression is throwing a parsing error in this case.

That's true for `["slice", 12345, 0, 1]`,

but not for `["slice", ['get', 'title'], 0, 1]` where `['get', 'title']` is a number:

https://jsbin.com/femayehebi/1/edit?html,output

In the latter case there's only a warning and it looks like the default value is used for `text-size` while in the former case the error prevents rendering the text label:

https://jsbin.com/rolodupela/1/edit?html,output

Is this behavior expected?

> Alternatively, the == expression could just pass on the sub-expression error.

:+1:
